README.md
# flutter_aliyun_captcha

适用于 Flutter 的阿里云人机验证插件

> 支持滑动验证、智能验证（智能验证暂不支持刮刮卡等形式）。

[![pub version][pub-image]][pub-url]

[pub-image]: https://img.shields.io/pub/v/flutter_aliyun_captcha.svg
[pub-url]: https://pub.dev/packages/flutter_aliyun_captcha

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [屏幕截图](#%E5%B1%8F%E5%B9%95%E6%88%AA%E5%9B%BE)
- [快速开始](#%E5%BF%AB%E9%80%9F%E5%BC%80%E5%A7%8B)
  - [安装](#%E5%AE%89%E8%A3%85)
  - [用法](#%E7%94%A8%E6%B3%95)
    - [滑动验证](#%E6%BB%91%E5%8A%A8%E9%AA%8C%E8%AF%81)
    - [智能验证](#%E6%99%BA%E8%83%BD%E9%AA%8C%E8%AF%81)
    - [获取 SDK 版本](#%E8%8E%B7%E5%8F%96-sdk-%E7%89%88%E6%9C%AC)
- [许可证](#%E8%AE%B8%E5%8F%AF%E8%AF%81)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## 屏幕截图

<div>
  <img src='https://github.com/leanflutter/flutter_aliyun_captcha/raw/master/screenshots/flutter_aliyun_captcha-ios-slide.png' width=280>
  <img src='https://github.com/leanflutter/flutter_aliyun_captcha/raw/master/screenshots/flutter_aliyun_captcha-ios-smart.png' width=280>
</div>

## 快速开始

### 安装

将此添加到包的 pubspec.yaml 文件中：

```yaml
dependencies:
  flutter_aliyun_captcha: ^1.0.5
```

您可以从命令行安装软件包：

```bash
$ flutter packages get
```

### 用法

导入 `flutter_aliyun_captcha`

```dart
import 'package:flutter_aliyun_captcha/flutter_aliyun_captcha.dart';
```

#### 滑动验证

> `AliyunCaptchaButton` 会根据其上层小部件的尺寸自适应，务必在上层小部件设置宽度和高度。

```dart
Container(
  width: double.infinity,
  height: 48,
  margin: EdgeInsets.only(
    top: 10,
    bottom: 10,
    left: 16,
    right: 16,
  ),
  child: AliyunCaptchaButton(
    controller: _captchaController,
    type: AliyunCaptchaType.slide, // 重要：请设置正确的类型
    option: AliyunCaptchaOption(
      appKey: '<appKey>',
      scene: 'scene',
      language: 'cn',
      // 更多参数请参见：https://help.aliyun.com/document_detail/193141.html
    ),
    customStyle: '''
      .nc_scale {
        background: #eeeeee !important;
        /* 默认背景色 */
      }

      .nc_scale div.nc_bg {
        background: #4696ec !important;
        /* 滑过时的背景色 */
      }

      .nc_scale .scale_text2 {
        color: #fff !important;
        /* 滑过时的字体颜色 */
      }

      .errloading {
        border: #ff0000 1px solid !important;
        color: #ef9f06 !important;
      }
    ''',
    onSuccess: (dynamic data) {
      // {"sig": "...", "token": "..."}
      _addLog('onSuccess', data);
    },
    onFailure: (String failCode) {
      _addLog('onFailure', 'failCode: $failCode');
    },
    onError: (String errorCode) {
      _addLog('onError', 'errorCode: $errorCode');
    },
  ),
)
```

#### 智能验证

```dart
AliyunCaptchaButton(
    controller: _captchaController,
    type: AliyunCaptchaType.smart, // 重要：请设置正确的类型
    option: AliyunCaptchaOption(
      appKey: '<appKey>',
      scene: 'scene',
      language: 'cn',
      // 更多参数请参见：https://help.aliyun.com/document_detail/193144.html
    ),
    onSuccess: (dynamic data) {
      // {"sig": "...", "token": "..."}
      _addLog('onSuccess', data);
    },
    onFailure: (String failCode) {
      _addLog('onFailure', 'failCode: $failCode');
    },
    onError: (String errorCode) {
      _addLog('onError', 'errorCode: $errorCode');
    },
  ),
)
```

#### 获取 SDK 版本

```dart
String sdkVersion = await AliyunCaptcha.sdkVersion;
```

## 许可证

```
MIT License

Copyright (c) 2021 LiJianying <lijy91@foxmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```


CHANGELOG.md
## 1.0.5

- fix android webview hardware #18

## 1.0.4

- 空安全支持
- [#9](https://github.com/leanflutter/flutter_aliyun_captcha/issues/9) 修复偶现滑块显示异常

## 1.0.3

- [#8](https://github.com/leanflutter/flutter_aliyun_captcha/issues/8) 增加 `customStyle` 属性以支持自定义样式

## 1.0.2

- [#7](https://github.com/leanflutter/flutter_aliyun_captcha/issues/7) 修复部分机型弹出安全键盘后验证按钮消失的问题

## 1.0.1

- 验证按钮宽度支持自适应

## 1.0.0

- 更新到新版人机验证（AWSC）
- 移除验证弹窗的调用形式，改为验证按钮

## 0.0.1

- 初始发行


example/lib/pages/home.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_aliyun_captcha/flutter_aliyun_captcha.dart';

const kLanguages = ['cn', 'en', 'ja_JP', 'ko_KR', 'ru_RU'];
const kTests = ['pass', 'block'];

class _ListSection extends StatelessWidget {
  final Widget? title;

  const _ListSection({
    Key? key,
    this.title,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.grey.withOpacity(0.1),
      padding: EdgeInsets.only(
        left: 16,
        right: 16,
        top: 10,
        bottom: 10,
      ),
      child: Column(
        children: [
          Row(
            children: [
              DefaultTextStyle(
                style: TextStyle(
                  color: Colors.black,
                  fontSize: 13,
                  fontWeight: FontWeight.bold,
                ),
                child: title!,
              ),
            ],
          ),
        ],
      ),
    );
  }
}

class _ListItem extends StatelessWidget {
  final Widget? title;
  final Widget? subtitle;
  final Widget? trailing;
  final VoidCallback? onTap;

  const _ListItem({
    Key? key,
    this.title,
    this.subtitle,
    this.trailing,
    this.onTap,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      behavior: HitTestBehavior.opaque,
      child: Container(
        constraints: BoxConstraints(minHeight: 48),
        padding: EdgeInsets.only(
          left: 16,
          right: 16,
          top: 8,
          bottom: 8,
        ),
        alignment: Alignment.centerLeft,
        child: Column(
          children: [
            Row(
              children: [
                DefaultTextStyle(
                  style: TextStyle(
                    color: Colors.black,
                    fontSize: 15,
                  ),
                  child: title!,
                ),
                Expanded(child: Container()),
                if (trailing != null) SizedBox(height: 34, child: trailing),
              ],
            ),
            if (subtitle != null) Container(child: subtitle),
          ],
        ),
      ),
      onTap: this.onTap,
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  AliyunCaptchaController _captchaController = AliyunCaptchaController();

  String _sdkVersion = 'Unknown';
  List<String> _logs = [];

  AliyunCaptchaType _captchaType = AliyunCaptchaType.slide;
  String _language = 'cn';
  String _test = '';

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String sdkVersion;
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      sdkVersion = await AliyunCaptcha.sdkVersion;
    } on PlatformException {
      sdkVersion = 'Failed to get platform version.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _sdkVersion = sdkVersion;
    });
  }

  void _addLog(String method, dynamic data) {
    _logs.add('>>>$method');
    if (data != null) {
      _logs.add(data is Map ? json.encode(data) : data);
    }
    _logs.add(' ');

    setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Plugin example app'),
      ),
      body: Container(
        child: Column(
          mainAxisSize: MainAxisSize.max,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: <Widget>[
            Container(
              alignment: Alignment.center,
              child: Column(
                children: <Widget>[
                  _ListItem(
                    title: Text('SDKVersion: $_sdkVersion'),
                    trailing: GestureDetector(
                      child: Container(
                        color: Theme.of(context).primaryColor,
                        padding: EdgeInsets.all(10),
                        child: Text(
                          'Reset',
                          style: TextStyle(
                            color: Colors.white,
                          ),
                        ),
                      ),
                      onTap: () {
                        _captchaController.reset();
                        _logs = [];
                        setState(() {});
                      },
                    ),
                  ),
                  Divider(height: 0, indent: 16, endIndent: 16),
                  _ListItem(
                    title: Text('captchaType'),
                    trailing: ToggleButtons(
                      children: <Widget>[
                        for (var captchaType in AliyunCaptchaType.values)
                          Text(captchaType.name),
                      ],
                      onPressed: (int index) {
                        _captchaType = AliyunCaptchaType.values[index];

                        setState(() {});
                      },
                      isSelected: AliyunCaptchaType.values
                          .map((e) => e == _captchaType)
                          .toList(),
                    ),
                  ),
                  _ListSection(
                    title: Text('Option'),
                  ),
                  _ListItem(
                    title: Text('language'),
                    trailing: ToggleButtons(
                      children: <Widget>[
                        for (var language in kLanguages) Text(language),
                      ],
                      onPressed: (int index) {
                        _language = kLanguages[index];

                        setState(() {});
                      },
                      isSelected:
                          kLanguages.map((e) => e == _language).toList(),
                    ),
                  ),
                  Divider(height: 0, indent: 16, endIndent: 16),
                  _ListItem(
                    title: Text('test'),
                    trailing: ToggleButtons(
                      children: <Widget>[
                        for (var test in kTests) Text(test),
                      ],
                      onPressed: (int index) {
                        _test = kTests[index];

                        setState(() {});
                      },
                      isSelected: kTests.map((e) => e == _test).toList(),
                    ),
                  ),
                  Divider(height: 0, indent: 16, endIndent: 16),
                  SizedBox(height: 10),
                  Container(
                    // width: double.infinity,
                    width: 280,
                    height: 44,
                    margin: EdgeInsets.only(
                      top: 10,
                      bottom: 10,
                      left: 16,
                      right: 16,
                    ),
                    child: AliyunCaptchaButton(
                      controller: _captchaController,
                      type: _captchaType,
                      option: AliyunCaptchaOption(
                        appKey: 'FFFF0N00000000006CAB',
                        scene: _captchaType == AliyunCaptchaType.slide
                            ? 'nc_login'
                            : 'ic_login',
                        language: _language,
                        // hideErrorCode: true,
                        test: _test != 'block'
                            ? null
                            : _captchaType == AliyunCaptchaType.slide
                                ? 'code300'
                                : 800,
                      ),
                      customStyle: '''
                        .nc_scale {
                          background: #eeeeee !important;
                          /* 默认背景色 */
                        }

                        .nc_scale div.nc_bg {
                          background: #4696ec !important;
                          /* 滑过时的背景色 */
                        }

                        .nc_scale .scale_text2 {
                          color: #fff !important;
                          /* 滑过时的字体颜色 */
                        }

                        .errloading {
                          border: #ff0000 1px solid !important;
                          color: #ef9f06 !important;
                        }
                      ''',
                      onSuccess: (dynamic data) {
                        // {"sig": "...", "token": "..."}
                        _addLog('onSuccess', data);
                      },
                      onFailure: (String failCode) {
                        _addLog('onFailure', 'failCode: $failCode');
                      },
                      onError: (String errorCode) {
                        _addLog('onError', 'errorCode: $errorCode');
                      },
                    ),
                  ),
                  SizedBox(height: 20),
                ],
              ),
            ),
            Expanded(
              child: SingleChildScrollView(
                padding: EdgeInsets.only(left: 16, right: 16, bottom: 20),
                child: Column(
                  mainAxisSize: MainAxisSize.max,
                  mainAxisAlignment: MainAxisAlignment.start,
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    for (var log in _logs) Text(log),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
